packages/common/src/converter/payload-converter.ts

```diff
@@ -140,3 +149,75 @@ export class BinaryPayloadConverter extends AsyncFacadePayloadConverter {
     return content.data as any;
   }
 }
+
+/**
+ * Converts between protobufjs Message instances and serialized Protobuf Payload
+ */
+export class ProtobufPayloadConverter extends AsyncFacadePayloadConverter {
+  public encodingType = encodingTypes.METADATA_ENCODING_PROTOBUF;
+
+  constructor(protected protobufClasses?: Record<string, unknown>) {
+    super();
+
+    if (protobufClasses && typeof protobufClasses !== 'object') {
+      throw new DataConverterError('protobufClasses must be an object');
+    }
+  }
+
+  public toDataSync(value: unknown): Payload | undefined {
+    const isProtobufMessageInstance =
+      this.protobufClasses && value && typeof value === 'object' && this.protobufClasses[value.constructor.name];
+    if (!isProtobufMessageInstance) {
+      return undefined;
+    }
+
+    // Safe to non-null assert due to above check
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const messageClass = this.validateMessageClass(this.protobufClasses![value.constructor.name]);
+
+    return {
+      metadata: {
+        [METADATA_ENCODING_KEY]: encodingKeys.METADATA_ENCODING_PROTOBUF,
+        [METADATA_MESSAGE_TYPE_KEY]: u8(value.constructor.name),
```

Comment:
Hrmm, not sure this should not be required as it should be the recipient who should choose what to deserialize into, not the serializer. This approach is also not cross-SDK compatible.

While this might seem reasonable, if someone wanted to do this, they could use `google.protobuf.Any` as their proto type which _does_ embed the fully-qualified message class in it.

My suggestion elsewhere about a generic `ProtobufMessage` result in `fromPayload` makes more sense now.

Reply:
> it should be the recipient who should choose what to deserialize into, not the serializer

Currently it's the sender who decides. Are there cases in which a user would send an instance of a certain protobuf type and want to receive a different type?

> This approach is also not cross-SDK compatible.

Correct—it would require modification of other SDKs to add the `METADATA_MESSAGE_TYPE_KEY` when encoding.

Reply:
> Currently it's the sender who decides.

It's both. Proto is just a binary blob. You can have different versions, different packages, etc. Same thing with JSON. The caller sends a JSON and the recipient can do what they want with it (deserialize into an existing object, sanitize in some way, etc).

> Are there cases in which a user would send an instance of a certain protobuf type and want to receive a different type?

Probably not

> Correct—it would require modification of other SDKs to add the `METADATA_MESSAGE_TYPE_KEY` when encoding.

Unfortunately the decision was made in this repository to not allow the user to specify the type in the data converter at conversion time. That decision should not require us to go add another metadata type to the other SDKs IMO.

Reply:
We should change the other SDKs to support the dynamic use case

Reply:
FWIW, they support this use case amongst each other. TS is the abnormal one that doesn't take the type during deserialization/decode.